File: src/main/java/com/NovumScientiaTeam/novumscientia/registry/NSMaterialRegistry.java
package com.NovumScientiaTeam.novumscientia.registry;

import com.EmosewaPixel.pixellib.materialsystem.element.Elements;
import com.EmosewaPixel.pixellib.materialsystem.materials.*;
import com.EmosewaPixel.pixellib.materialsystem.materials.utility.GroupMaterial;
import com.EmosewaPixel.pixellib.materialsystem.materials.utility.TransitionMaterial;

import static com.EmosewaPixel.pixellib.materialsystem.MaterialRegistry.*;

public class NSMaterialRegistry {
    //Material Tags
    public static final String DECOMPOSABLE = "decomposable";
    public static final String DISABLE_FACTORY_COMPONENTS = "disable_factory_components";
    public static final String NOT_REACTIVE = "not_reactive";
    public static final String SUPER_ORE = "super_ore";
    public static final String IS_ACID = "is_acid";

    //Element Materials
    public static Material HYDROGEN, HYDROGEN_GAS, DEUTERIUM, DEUTERIUM_GAS, TRITIUM, TRITIUM_GAS, HELIUM, HELIUM_3, LITHIUM, BERYLLIUM, BORON, CARBON, NITROGEN, NITROGEN_GAS, OXYGEN, OXYGEN_GAS, FLUORINE, FLUORINE_GAS, NEON, SODIUM, MAGNESIUM, ALUMINIUM, SILICON, PHOSPHORUS, SULFUR, CHLORINE, CHLORINE_GAS, ARGON, POTASSIUM, CALCIUM, SCANDIUM, TITANIUM, VANADIUM, CHROMIUM, MANGANESE, COBALT, NICKEL, COPPER, ZINC, GALLIUM, GERMANIUM, ARSENIC, BROMINE, BROMINE_GAS, KRYPTON;

    //Compound Materials
    public static Material SULFUR_DIOXIDE, SILICON_MONOXIDE;

    //Group Materials
    public static Material HYDROXIDE, SULFATE, PHOSPHATE, METASILICATE;

    //Ore Materials
    public static Material MAGNETITE, HEMATITE, PYRITE, CHALCOPYRITE, CHALCOCITE, CUPRITE, TETRAHEDRITE, ROCK_SALT, SYLVITE, KAOLINITE;


    //Alloy Materials


    static {
        HYDROGEN = new TransitionMaterial("hydrogen", () -> HYDROGEN_GAS, 2).setElement(Elements.HYDROGEN).build();
        HYDROGEN_GAS = new FluidMaterial("hydrogen_has", TRANSPARENT_FLUID, 0x31219e).addTags(IS_GAS).setComposition(new MaterialStack(HYDROGEN, 2)).build();
        DEUTERIUM = new TransitionMaterial("deuterium", () -> DEUTERIUM_GAS, 2).setElement(Elements.DEUTERIUM).build();
        DEUTERIUM_GAS = new FluidMaterial("deuterium_gas", TRANSPARENT_FLUID, 0xfffdbf).addTags(IS_GAS).setComposition(new MaterialStack(DEUTERIUM, 2)).build();
        TRITIUM = new TransitionMaterial("tritium", () -> TRITIUM_GAS, 2).setElement(Elements.DEUTERIUM).build();
        TRITIUM_GAS = new FluidMaterial("tritium_gas", TRANSPARENT_FLUID, 0xfffc9e).addTags(IS_GAS).setComposition(new MaterialStack(TRITIUM, 2)).build();
        HELIUM = new FluidMaterial("helium", TRANSPARENT_FLUID, 0xeebe21).addTags(IS_GAS).setElement(Elements.HELIUM).build();
        HELIUM_3 = new FluidMaterial("helium_3", TRANSPARENT_FLUID, 0xf60d0b).addTags(IS_GAS).setElement(Elements.HELIUM_3).build();
        LITHIUM = new IngotMaterial("lithium", REGULAR, 0xa9afac, 1).setElement(Elements.LITHIUM).setMeltingTemperature(453).setBoilingTemperature(1603).build();
//        BERYLLIUM = new
        BORON = new IngotMaterial("boron", SHINY, 0x727272, 1).setElement(Elements.BORON).setMeltingTemperature(2349).setBoilingTemperature(4200).build();
        CARBON = new IngotMaterial("carbon", REGULAR, 0x474747 , 1).setElement(Elements.CARBON).setMeltingTemperature(3915).setBoilingTemperature(3915).build();
        NITROGEN = new TransitionMaterial("nitrogen", () -> NITROGEN_GAS, 2).setElement(Elements.NITROGEN).build();
        NITROGEN_GAS = new FluidMaterial("nitrogen_gas", TRANSPARENT_FLUID, 0x8f8fff).addTags(IS_GAS).setElement(Elements.NITROGEN).build();
        OXYGEN = new TransitionMaterial("oxygen", () -> OXYGEN_GAS, 2).setElement(Elements.OXYGEN).build();
        OXYGEN_GAS = new FluidMaterial("oxygen_gas", TRANSPARENT_FLUID, 0x56badd).addTags(IS_GAS).setComposition(new MaterialStack(OXYGEN, 2)).build();
        FLUORINE = new TransitionMaterial("fluorine", () -> FLUORINE_GAS, 2).setElement(Elements.FLUORINE).build();
        FLUORINE_GAS = new FluidMaterial("fluorine_gas", OPAQUE_FLUID, 0xdee67f).addTags(IS_GAS).setElement(Elements.FLUORINE).build();
        NEON = new FluidMaterial("neon", TRANSPARENT_FLUID, 0xb3e3f5).addTags(IS_GAS).setElement(Elements.NEON).build();
        SODIUM = new IngotMaterial("sodium", SHINY, 0x525252, 1).setElement(Elements.SODIUM).setMeltingTemperature(371).setBoilingTemperature(1156).build();
        MAGNESIUM = new IngotMaterial("magnesium", SHINY, 0x8a8a8a,1).setElement(Elements.MAGNESIUM).setMeltingTemperature(923).setBoilingTemperature(1363).build();
        ALUMINIUM = new IngotMaterial("aluminium", REGULAR, 0x96a4b0, 1).setSecondName("aluminum").setElement(Elements.ALUMINUM).setMeltingTemperature(933).setBoilingTemperature(2743).build();
        SILICON = new IngotMaterial("silicon", REGULAR, 0x121315, 1).setElement(Elements.SILICON).setMeltingTemperature(1687).setBoilingTemperature(3538).build();
        PHOSPHORUS = new DustMaterial("phosphorus", FINE, 0xf50b0b, 1).setElement(Elements.PHOSPHORUS).build();
        SULFUR = new DustMaterial("sulfur", FINE, 0xfff70c, 1).setElement(Elements.SULFUR).build();

        COPPER = new IngotMaterial("copper", SHINY, 0xca7966, 1).setElement(Elements.COPPER).setMeltingTemperature(1358).setBoilingTemperature(2835).build();

        SULFUR_DIOXIDE = new FluidMaterial("sulfur_dioxide", TRANSPARENT_FLUID, 0xf8d34d).addTags(IS_GAS, DECOMPOSABLE).setComposition(new MaterialStack(SULFUR), new MaterialStack(OXYGEN_GAS)).build();
        SILICON_MONOXIDE = new DustMaterial("silicon_monoxide", FINE, 0x2b2b2a, 1).addTags(DECOMPOSABLE).setComposition(new MaterialStack(SILICON), new MaterialStack(OXYGEN)).build();


        HYDROXIDE = new GroupMaterial("hydroxide").setComposition(new MaterialStack(HYDROGEN), new MaterialStack(OXYGEN)).build();
        SULFATE = new GroupMaterial("sulfate").setComposition(new MaterialStack(SULFUR_DIOXIDE), new MaterialStack(OXYGEN_GAS)).build();
        PHOSPHATE = new GroupMaterial("phosphate").setComposition(new MaterialStack(PHOSPHORUS), new MaterialStack(OXYGEN_GAS, 2)).build();
        METASILICATE = new GroupMaterial("metasilicate").setComposition(new MaterialStack(SILICON_MONOXIDE), new MaterialStack(OXYGEN_GAS)).build();


        MAGNETITE = new DustMaterial("magnetite", FINE, 0x1f1e1c, 1).addTags(HAS_ORE).setRefinedMaterial((IngotMaterial) IRON).setComposition(new MaterialStack(IRON, 3), new MaterialStack(OXYGEN_GAS, 2)).build();
        HEMATITE = new DustMaterial("hematite", FINE, 0x444138, 1).addTags(HAS_ORE).setRefinedMaterial((IngotMaterial) IRON).setComposition(new MaterialStack(IRON, 2), new MaterialStack(OXYGEN, 3)).build();
        PYRITE = new DustMaterial("pyrite", FINE, 0xc0a368, 1).addTags(HAS_ORE).setRefinedMaterial((IngotMaterial) IRON).setComposition(new MaterialStack(IRON), new MaterialStack(SULFUR, 2)).build();
        CHALCOPYRITE = new DustMaterial("chalcopyrite", FINE, 0x9d8851, 1).addTags(HAS_ORE).setRefinedMaterial((IngotMaterial) COPPER).setComposition(new MaterialStack(COPPER), new MaterialStack(IRON), new MaterialStack(SULFUR)).build();
        CHALCOCITE = new DustMaterial("chalcocite", FINE, 0xa1ba81, 1).addTags(HAS_ORE).setRefinedMaterial((IngotMaterial) COPPER).setComposition(new MaterialStack(COPPER, 2), new MaterialStack(SULFUR)).build();
        CUPRITE = new DustMaterial("cuprite", FINE, 0x458a77, 1).addTags(HAS_ORE).setRefinedMaterial((IngotMaterial) COPPER).setComposition(new MaterialStack(COPPER, 2), new MaterialStack(OXYGEN)).build();

        KAOLINITE = new DustMaterial("kaolinite", FINE, 0xe7c9c0, 1).setComposition(new MaterialStack(ALUMINIUM, 2), new MaterialStack(SILICON, 2), new MaterialStack(OXYGEN, 5), new MaterialStack(HYDROXIDE, 4)).addTags(DECOMPOSABLE).build();

        BRICK.addTags(DISABLE_FACTORY_COMPONENTS);
        NETHER_BRICK.addTags(DISABLE_FACTORY_COMPONENTS);
        DIAMOND.addTags(NOT_REACTIVE);
        WATER.setComposition(new MaterialStack(HYDROGEN_GAS), new MaterialStack(OXYGEN));
        QUARTZ.setComposition(new MaterialStack(SILICON), new MaterialStack(OXYGEN_GAS));
    }
}

File: src/main/java/com/NovumScientiaTeam/novumscientia/registry/NSMaterialRegistry.kt
package com.novumscientiateam.novumscientia.registry

import com.emosewapixel.pixellib.materialsystem.*
import com.emosewapixel.pixellib.materialsystem.MaterialRegistry.HAS_ORE
import com.emosewapixel.pixellib.materialsystem.MaterialRegistry.IS_GAS
import com.emosewapixel.pixellib.materialsystem.element.Elements
import com.emosewapixel.pixellib.materialsystem.materials.IngotMaterial
import com.emosewapixel.pixellib.materialsystem.materials.Material
import com.emosewapixel.pixellib.materialsystem.materials.utility.MaterialStack

object NSMaterialRegistry {
    //Material Tags
    const val DECOMPOSABLE = "decomposable"
    const val DISABLE_FACTORY_COMPONENTS = "disable_factory_components"
    const val NOT_REACTIVE = "not_reactive"
    const val SUPER_ORE = "super_ore"
    const val IS_ACID = "is_acid"

    //Element Materials
    @JvmField
    val HYDROGEN: Material
    @JvmField
    var HYDROGEN_GAS: Material? = null
    @JvmField
    val DEUTERIUM: Material
    @JvmField
    var DEUTERIUM_GAS: Material? = null
    @JvmField
    val TRITIUM: Material
    @JvmField
    var TRITIUM_GAS: Material? = null
    @JvmField
    val HELIUM: Material
    @JvmField
    val HELIUM_3: Material
    @JvmField
    val LITHIUM: Material
    //@JvmField
    //val BERYLLIUM: Material
    @JvmField
    val BORON: Material
    @JvmField
    val CARBON: Material
    @JvmField
    val NITROGEN: Material
    @JvmField
    var NITROGEN_GAS: Material? = null
    @JvmField
    val OXYGEN: Material
    @JvmField
    var OXYGEN_GAS: Material? = null
    @JvmField
    val FLUORINE: Material
    @JvmField
    var FLUORINE_GAS: Material? = null
    @JvmField
    val NEON: Material
    @JvmField
    val SODIUM: Material
    @JvmField
    val MAGNESIUM: Material
    @JvmField
    val ALUMINIUM: Material
    @JvmField
    val SILICON: Material
    @JvmField
    val PHOSPHORUS: Material
    @JvmField
    val SULFUR: Material
    //@JvmField
    //val CHLORINE: Material
    //@JvmField
    //val CHLORINE_GAS: Material
    //@JvmField
    //val ARGON: Material
    //@JvmField
    //val POTASSIUM: Material
    //@JvmField
    //val CALCIUM: Material
    //@JvmField
    //val SCANDIUM: Material
    //@JvmField
    //val TITANIUM: Material
    //@JvmField
    //val VANADIUM: Material
    //@JvmField
    //val CHROMIUM: Material
    //@JvmField
    //val MANGANESE: Material
    //@JvmField
    //val COBALT: Material
    //@JvmField
    //val NICKEL: Material
    @JvmField
    val COPPER: Material
    //@JvmField
    //val ZINC: Material
    //@JvmField
    //val GALLIUM: Material
    //@JvmField
    //val GERMANIUM: Material
    //@JvmField
    //val ARSENIC: Material
    //@JvmField
    //val BROMINE: Material
    //@JvmField
    //val BROMINE_GAS: Material
    //@JvmField
    //val KRYPTON: Material
    //@JvmField

    //Compound Materials
    @JvmField
    val SULFUR_DIOXIDE: Material
    @JvmField
    val SILICON_MONOXIDE: Material

    //Group Materials
    @JvmField
    val HYDROXIDE: Material
    @JvmField
    val SULFATE: Material
    @JvmField
    val PHOSPHATE: Material
    @JvmField
    val METASILICATE: Material

    //Ore Materials
    @JvmField
    val MAGNETITE: Material
    @JvmField
    val HEMATITE: Material
    @JvmField
    val PYRITE: Material
    @JvmField
    val CHALCOPYRITE: Material
    @JvmField
    val CHALCOCITE: Material
    @JvmField
    val CUPRITE: Material
    //@JvmField
    //val TETRAHEDRITE: Material
    //@JvmField
    //val ROCK_SALT: Material
    //@JvmField
    //val SYLVITE: Material
    @JvmField
    val KAOLINITE: Material

    //Alloy Materials


    init {
        HYDROGEN = transitionMaterial("hydrogen", { HYDROGEN_GAS!! }, 2) {
            element = Elements.HYDROGEN
        }
        HYDROGEN_GAS = fluidMaterial("hydrogen_gas", MaterialRegistry.TRANSPARENT_FLUID, 0x31219e) {
            materialTags += IS_GAS
            composition = listOf(MaterialStack(HYDROGEN, 2))
        }
        DEUTERIUM = transitionMaterial("deuterium", { DEUTERIUM_GAS!! }, 2) {
            element = Elements.DEUTERIUM
        }
        DEUTERIUM_GAS = fluidMaterial("deuterium_gas", MaterialRegistry.TRANSPARENT_FLUID, 0xfffdbf) {
            materialTags += IS_GAS
            composition = listOf(MaterialStack(DEUTERIUM, 2))
        }
        TRITIUM = transitionMaterial("tritium", { TRITIUM_GAS!! }, 2) {
            element = Elements.DEUTERIUM
        }
        TRITIUM_GAS = fluidMaterial("tritium_gas", MaterialRegistry.TRANSPARENT_FLUID, 0xfffc9e) {
            materialTags += IS_GAS
            composition = listOf(MaterialStack(TRITIUM, 2))
        }
        HELIUM = fluidMaterial("helium", MaterialRegistry.TRANSPARENT_FLUID, 0xeebe21) {
            materialTags += IS_GAS
            element = Elements.HELIUM
        }
        HELIUM_3 = fluidMaterial("helium_3", MaterialRegistry.TRANSPARENT_FLUID, 0xf60d0b) {
            materialTags += IS_GAS
            element = Elements.HELIUM_3
        }
        LITHIUM = ingotMaterial("lithium", MaterialRegistry.REGULAR, 0xa9afac, 1) {
            element = Elements.LITHIUM
            meltingTemperature = 453
            boilingTemperature = 1603
        }

        BORON = ingotMaterial("boron", MaterialRegistry.SHINY, 0x727272, 1) {
            element = Elements.BORON
            meltingTemperature = 2349
            boilingTemperature = 4200
        }
        CARBON = ingotMaterial("carbon", MaterialRegistry.REGULAR, 0x474747, 1) {
            element = Elements.CARBON
            meltingTemperature = 3915
            boilingTemperature = 3915
        }
        NITROGEN = transitionMaterial("nitrogen", { NITROGEN_GAS!! }, 2) {
            element = Elements.NITROGEN
        }
        NITROGEN_GAS = fluidMaterial("nitrogen_gas", MaterialRegistry.TRANSPARENT_FLUID, 0x8f8fff) {
            materialTags += IS_GAS
            element = Elements.NITROGEN
        }
        OXYGEN = transitionMaterial("oxygen", { OXYGEN_GAS!! }, 2) {
            element = Elements.OXYGEN
        }
        OXYGEN_GAS = fluidMaterial("oxygen_gas", MaterialRegistry.TRANSPARENT_FLUID, 0x56badd) {
            materialTags += IS_GAS
            composition = listOf(MaterialStack(OXYGEN, 2))
        }
        FLUORINE = transitionMaterial("fluorine", { FLUORINE_GAS!! }, 2) {
            element = Elements.FLUORINE
        }
        FLUORINE_GAS = fluidMaterial("fluorine_gas", MaterialRegistry.OPAQUE_FLUID, 0xdee67f) {
            materialTags += IS_GAS
            element = Elements.FLUORINE
        }
        NEON = fluidMaterial("neon", MaterialRegistry.TRANSPARENT_FLUID, 0xb3e3f5) {
            materialTags += IS_GAS
            element = Elements.NEON
        }
        SODIUM = ingotMaterial("sodium", MaterialRegistry.SHINY, 0x525252, 1) {
            element = Elements.SODIUM
            meltingTemperature = 371
            boilingTemperature = 1156
        }
        MAGNESIUM = ingotMaterial("magnesium", MaterialRegistry.SHINY, 0x8a8a8a, 1) {
            element = Elements.MAGNESIUM
            meltingTemperature = 923
            boilingTemperature = 1363
        }
        ALUMINIUM = ingotMaterial("aluminium", MaterialRegistry.REGULAR, 0x96a4b0, 1) {
            secondName = "aluminum"
            element = Elements.ALUMINUM
            meltingTemperature = 933
            boilingTemperature = 2743
        }
        SILICON = ingotMaterial("silicon", MaterialRegistry.REGULAR, 0x121315, 1) {
            element = Elements.SILICON
            meltingTemperature = 1687
            boilingTemperature = 3538
        }
        PHOSPHORUS = dustMaterial("phosphorus", MaterialRegistry.FINE, 0xf50b0b, 1) {
            element = Elements.PHOSPHORUS
        }
        SULFUR = dustMaterial("sulfur", MaterialRegistry.FINE, 0xfff70c, 1) {
            element = Elements.SULFUR
        }

        COPPER = ingotMaterial("copper", MaterialRegistry.SHINY, 0xca7966, 1) {
            element = Elements.COPPER
            meltingTemperature = 1358
            boilingTemperature = 2835
        }


        SULFUR_DIOXIDE = fluidMaterial("sulfur_dioxide", MaterialRegistry.TRANSPARENT_FLUID, 0xf8d34d) {
            materialTags += listOf(IS_GAS, DECOMPOSABLE)
            composition = listOf(MaterialStack(SULFUR), MaterialStack(OXYGEN_GAS!!))
        }
        SILICON_MONOXIDE = dustMaterial("silicon_monoxide", MaterialRegistry.FINE, 0x2b2b2a, 1) {
            materialTags += DECOMPOSABLE
            composition = listOf(MaterialStack(SILICON), MaterialStack(OXYGEN))
        }


        HYDROXIDE = groupMaterial("hydroxide") {
            composition = listOf(MaterialStack(HYDROGEN), MaterialStack(OXYGEN))
        }
        SULFATE = groupMaterial("sulfate") {
            composition = listOf(MaterialStack(SULFUR_DIOXIDE), MaterialStack(OXYGEN_GAS!!))
        }
        PHOSPHATE = groupMaterial("phosphate") {
            composition = listOf(MaterialStack(PHOSPHORUS), MaterialStack(OXYGEN_GAS!!, 2))
        }
        METASILICATE = groupMaterial("metasilicate") {
            composition = listOf(MaterialStack(SILICON_MONOXIDE), MaterialStack(OXYGEN_GAS!!))
        }


        MAGNETITE = dustMaterial("magnetite", MaterialRegistry.FINE, 0x1f1e1c, 1) {
            materialTags += HAS_ORE
            refinedMaterial = MaterialRegistry.IRON as IngotMaterial
            composition = listOf(MaterialStack(MaterialRegistry.IRON, 3), MaterialStack(OXYGEN_GAS!!, 2))
        }
        HEMATITE = dustMaterial("hematite", MaterialRegistry.FINE, 0x444138, 1) {
            materialTags += HAS_ORE
            refinedMaterial = MaterialRegistry.IRON as IngotMaterial
            composition = listOf(MaterialStack(MaterialRegistry.IRON, 2), MaterialStack(OXYGEN, 3))
        }
        PYRITE = dustMaterial("pyrite", MaterialRegistry.FINE, 0xc0a368, 1) {
            materialTags += HAS_ORE
            refinedMaterial = MaterialRegistry.IRON as IngotMaterial
            composition = listOf(MaterialStack(MaterialRegistry.IRON), MaterialStack(SULFUR, 2))
        }
        CHALCOPYRITE = dustMaterial("chalcopyrite", MaterialRegistry.FINE, 0x9d8851, 1) {
            materialTags += HAS_ORE
            refinedMaterial = COPPER as IngotMaterial
            composition = listOf(MaterialStack(COPPER), MaterialStack(MaterialRegistry.IRON), MaterialStack(SULFUR))
        }
        CHALCOCITE = dustMaterial("chalcocite", MaterialRegistry.FINE, 0xa1ba81, 1) {
            materialTags += HAS_ORE
            refinedMaterial = COPPER as IngotMaterial
            composition = listOf(MaterialStack(COPPER, 2), MaterialStack(SULFUR))
        }
        CUPRITE = dustMaterial("cuprite", MaterialRegistry.FINE, 0x458a77, 1) {
            materialTags += HAS_ORE
            refinedMaterial = COPPER as IngotMaterial
            composition = listOf(MaterialStack(COPPER, 2), MaterialStack(OXYGEN))
        }


        KAOLINITE = dustMaterial("kaolinite", MaterialRegistry.FINE, 0xe7c9c0, 1) {
            composition = listOf(MaterialStack(ALUMINIUM, 2), MaterialStack(SILICON, 2), MaterialStack(OXYGEN, 5), MaterialStack(HYDROXIDE, 4))
            materialTags += DECOMPOSABLE
        }


        MaterialRegistry.BRICK {
            materialTags += DISABLE_FACTORY_COMPONENTS
        }
        MaterialRegistry.NETHER_BRICK {
            materialTags += DISABLE_FACTORY_COMPONENTS
        }
        MaterialRegistry.DIAMOND {
            materialTags += NOT_REACTIVE
        }
        MaterialRegistry.WATER {
            composition = listOf(MaterialStack(HYDROGEN_GAS!!), MaterialStack(OXYGEN))
        }
        MaterialRegistry.QUARTZ {
            composition = listOf(MaterialStack(SILICON), MaterialStack(OXYGEN_GAS!!))
        }
    }
}